flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptivebatch/BatchJobRecoveryTest.java

```diff
@@ -0,0 +1,1217 @@
+
+        // the second time trigger partition not found
+        firstMiddle0 = getExecutionVertex(MIDDLE_ID, 0, newScheduler.getExecutionGraph());
+        triggerFailedByDataConsumptionException(newScheduler, firstMiddle0);
+
+        waitUntilExecutionVertexState(firstMiddle0, ExecutionState.CREATED, 5000L);
+
+        // verify only source task 0 were restarted
+        assertThat(
+                        getExecutionVertex(SOURCE_ID, 0, newScheduler.getExecutionGraph())
+                                .getExecutionState())
+                .isNotEqualTo(ExecutionState.FINISHED);
+
+        for (int i = 1; i < 5; i++) {
+            assertThat(
+                            getExecutionVertex(SOURCE_ID, i, newScheduler.getExecutionGraph())
+                                    .getExecutionState())
+                    .isEqualTo(ExecutionState.FINISHED);
+        }
+    }
+
+    @TestTemplate
+    void testReplayEventFailed() throws Exception {
+        final JobEventStore failingJobEventStore =
+                new JobEventStore() {
+                    @Override
+                    public void start() {}
+
+                    @Override
+                    public void stop(boolean clear) {}
+
+                    @Override
+                    public void writeEvent(JobEvent event, boolean cutBlock) {}
+
+                    @Override
+                    public JobEvent readEvent() throws Exception {
+                        throw new Exception();
+                    }
+
+                    @Override
+                    public boolean isEmpty() {
+                        return false;
+                    }
+                };
+
+        final ManuallyTriggeredScheduledExecutor taskRestartExecutor =
+                new ManuallyTriggeredScheduledExecutor();
+        delayedExecutor = taskRestartExecutor;
+
+        final AdaptiveBatchScheduler newScheduler =
+                createScheduler(deserializeJobGraph(serializedJobGraph), failingJobEventStore);
+        startSchedulingAndWaitRecoverFinish(newScheduler);
+
+        // trigger scheduled restarting and drain the main thread actions
+        taskRestartExecutor.triggerScheduledTasks();
+        runInMainThread(() -> {});
+
+        assertThat(
+                        ExceptionUtils.findThrowableWithMessage(
+                                newScheduler.getExecutionGraph().getFailureCause(),
+                                "Recover failed from JM failover"))
+                .isPresent();
+
+        // source should be scheduled.
+        for (ExecutionVertex vertex :
+                getExecutionVertices(SOURCE_ID, newScheduler.getExecutionGraph())) {
+            assertThat(vertex.getCurrentExecutionAttempt().getAttemptNumber()).isEqualTo(1);
+            assertThat(vertex.getExecutionState()).isEqualTo(ExecutionState.DEPLOYING);
+        }
+    }
+
+    @TestTemplate
+    void testRecoverFromJMFailover() throws Exception {
```

Comment:
I think this is the most common case and should be the first case in this test set. Other cases just cover some special cases.